woocommerce/woocommerce-android
# Fix for Memory Leak in observeNotificationsPermissionBarVisibility Method

### Description


This PR addresses a memory leak issue found in the `observeNotificationsPermissionBarVisibility` method. The source of the leak was the Handler().postDelayed() method which retained a reference to the enclosing class until the delayed operation was executed.

The code is modified to hold onto the Handler and Runnable as member references. This approach allows us to explicitly remove the callbacks in the lifecycle onDestroy method.

Thanks @malinajirka, for the pair programming session.

### Testing instructions


In the MainActivity, change the `NOTIFICATIONS_PERMISSION_BAR_DISPLAY_DELAY` to 7000L 
Run the app
In the My Store Screen, rotate the screen once or twice and wait for the Notification Permission Banner to be displayed. 
Make sure Leak Canary does not flag any leaks. 

Here's the leak trace in case it's needed:

```
┬───
│ GC Root: System class
│
├─ android.app.ActivityThread class
│    Leaking: NO (MessageQueue↓ is not leaking and a class is never leaking)
│    ↓ static ActivityThread.sMainThreadHandler
├─ android.app.ActivityThread$H instance
│    Leaking: NO (MessageQueue↓ is not leaking)
│    ↓ Handler.mQueue
├─ android.os.MessageQueue instance
│    Leaking: NO (MessageQueue#mQuitting is false)
│    HandlerThread: "main"
│    ↓ MessageQueue[128]
│                  ~~~~~
├─ android.os.Message instance
│    Leaking: UNKNOWN
│    Retaining 1.5 MB in 29984 objects
│    Message.what = 0
│    Message.when = 6534517 (18625 ms after heap dump)
│    Message.obj = null
│    Message.callback = instance @342968120 of com.woocommerce.android.ui.main.
│    MainActivity$observeNotificationsPermissionBarVisibility$1$$ExternalSynthet
│    icLambda0
│    Message.target = instance @342968200 of android.os.Handler
│    ↓ Message.callback
│              ~~~~~~~~
├─ com.woocommerce.android.ui.main.
│  MainActivity$observeNotificationsPermissionBarVisibility$1$$ExternalSynthetic
│  Lambda0 instance
│    Leaking: UNKNOWN
│    Retaining 1.5 MB in 29982 objects
│    f$0 instance of com.woocommerce.android.ui.main.MainActivity with
│    mDestroyed = true
│    ↓
│    MainActivity$observeNotificationsPermissionBarVisibility$1$$ExternalSynthet
│    icLambda0.f$0
│              ~~~
╰→ com.woocommerce.android.ui.main.MainActivity instance
​     Leaking: YES (ObjectWatcher was watching this because com.woocommerce.
​     android.ui.main.MainActivity received Activity#onDestroy() callback and
​     Activity#mDestroyed is true)
​     Retaining 1.5 MB in 29981 objects
​     key = de2b1204-e2bd-45a4-8564-287387807467
​     watchDurationMillis = 5816
​     retainedDurationMillis = 644
​     mApplication instance of com.woocommerce.android.WooCommerceDebug
​     mBase instance of androidx.appcompat.view.ContextThemeWrapper

METADATA

Build.VERSION.SDK_INT: 33
Build.MANUFACTURER: Google
LeakCanary version: 2.9.1
App process name: com.woocommerce.android.dev
Class count: 38958
Instance count: 433823
Primitive array count: 205683
Object array count: 64003
Thread count: 133
Heap total bytes: 47339676
Bitmap count: 26
Bitmap total bytes: 489093
Large bitmap count: 0
Large bitmap total bytes: 0
Db 1: open /data/user/0/com.woocommerce.android.dev/databases/tracks.db
Db 2: open /data/user/0/com.woocommerce.android.dev/databases/com.google.
android.datatransport.events
Db 3: open /data/user/0/com.woocommerce.android.
dev/databases/wc-android-database
Db 4: open /data/user/0/com.woocommerce.android.
dev/databases/wp-android-database
Db 5: closed /data/user/0/com.woocommerce.android.
dev/databases/google_app_measurement_local.db
Db 6: open /data/user/0/com.woocommerce.android.dev/databases/wp-fluxc
Db 7: open /data/user/0/com.woocommerce.android.dev/no_backup/androidx.work.
workdb
Stats: LruCache[maxSize=3000,hits=176714,misses=380516,hitRate=31%]
RandomAccess[bytes=20310114,reads=380516,travel=188883854116,range=57456681,size
=69443785]
Analysis duration: 15514 ms
```
